transfers/reingest.py

```diff
@@ -0,0 +1,509 @@
+
+
+def compare_lists(user_list, system_list):
+    """If we have a user list we can compare it against the system list.
+    This helps, for one, to make sure our user list is correct. It also
+    helps us to test this script as a mechanism for identifying compressed
+    AIPs in future.
+    """
+    if set(user_list).issubset(set(system_list)):
+        LOGGER.info("User list is a subset of system list.")
+        return True
+
+
+def reingest_full_and_approve(amclient, pipeline_uuid, aip_uuid,
+                              processing_config="default", latency=None,
+                              approval_retries=2):
+    """Reingest the AIP. This function will make three calls to the AM Client
+    code. First to initiate the reingest, second to monitor the transfer status
+    if the reingest can begin. Third to approve it in the transfer workflow.
+    Providing a latency argument makes up for the endpoint being neither
+    particularly synchronous or asynchronous. We don't want to mindlessly poll
+    the server to check that the AIP we want to reingest is in the pipeline.
+    """
+
+    # Initialize reingest for an AIP.
+    reingest_aip = amclient.reingest_aip(pipeline_uuid=pipeline_uuid,
+                                         aip_uuid=aip_uuid,
+                                         processing_config=processing_config)
+
+    # If we successfully initialize the reingest we get a dict back that
+    # we can then work with to monitor state. If we do not get a dict back
+    # we need to look at what went wrong.
+    if not isinstance(reingest_aip, dict):
+        LOGGER.error("Reingest failed with server error %s, returning.",
+                     errors.error_lookup(reingest_aip))
+        return False, "Error calling reingest_aip."
+
+    reingest_uuid = reingest_aip["reingest_uuid"]
+    LOGGER.info("Reingest UUID to work with %s", reingest_uuid)
+
+    LOGGER.info("Checking status of %s once in transfer queue", reingest_uuid)
+
+    # We need to make sure that the approval is synchronized with the request
+    # to reingest. Occasionally a reingest will get stopped at approval when
+    # it doesn't need to if we just wait a little longer or try again.
+    for retry in range(approval_retries):
+        transfer = None
+        while not isinstance(transfer, dict):
+            if latency is not None and isinstance(latency, float):
+                time.sleep(latency)   # ~latency between call and AM actioning.
+            transfer = amclient.get_transfer_status(reingest_uuid)
+
+        LOGGER.info("Attempting to approve transfer following reingest init")
+
+        if transfer["status"] == "USER_INPUT":
+            transfer_directory = transfer["directory"]
+            LOGGER.info("Approving reingest automatically. Directory to "
+                        "approve %s", transfer_directory)
+            message = amclient.approve_transfer(transfer_directory)
+            if 'Approval successful.' in message['message']:
```

Comment:
I think a more robust test for a successful approval would be `if message.get('error') is None:` because a successful call will always return a dict with no `'error'` key (if I'm reading https://github.com/artefactual/archivematica/blob/stable/1.7.x/src/dashboard/src/components/api/views.py#L364 right)

Reply:
Good spot!